rust/protocol/src/sealed_sender.rs

```diff
@@ -879,8 +1019,214 @@ mod sealed_sender_v2 {
+            &sender_c_0,
+        )?;
+        assert_eq!(recv_at_0.as_ref(), sender_at_0.as_ref());
+
+        Ok(())
+    }
 }
 
+/// This method implements a single-key multi-recipient [KEM] as defined in Manuel Barbosa's
+/// ["Randomness Reuse: Extensions and Improvements"], a.k.a. Sealed Sender v2.
+///
+/// [KEM]: https://en.wikipedia.org/wiki/Key_encapsulation
+/// ["Randomness Reuse: Extensions and Improvements"]: https://haslab.uminho.pt/mbb/files/reuse.pdf
+///
+/// # Contrast with Sealed Sender v1
+/// The KEM scheme implemented by this method uses the "Generic Construction" in `4.1` of [that
+/// paper]["Randomness Reuse: Extensions and Improvements"], instantiated with [ElGamal
+/// encryption]. This technique enables reusing a single sequence of random bytes across multiple
+/// messages with the same content, which reduces computation time for clients sending the same
+/// message to multiple recipients (without compromising the message security).
+///
+/// There are a few additional design tradeoffs this method makes vs [Sealed Sender v1]
+/// which may make it comparatively unwieldy for certain scenarios:
+/// 1. it requires a [SessionRecord] to exist already for the recipient, i.e. that a Double Ratchet
+///    message chain has previously been established in the [SessionStore] via
+///    [crate::process_prekey_bundle] after an initial [crate::PreKeySignalMessage] is recieved.
+/// 2. it ferries a lot of additional information in its encoding which makes the resulting message
+///    bulkier than the message produced by this method.
+/// 3. unlike other message types sent over the wire, the encoded message returned by this method
+///    does *not* correspond to a specific protobuf schema, in order to avoid inefficiencies
+///    produced by protobuf's packing (see **[Wire Format]**).
+///
+/// [ElGamal encryption]: https://en.wikipedia.org/wiki/ElGamal_encryption
+/// [Sealed Sender v1]: sealed_sender_encrypt_from_usmc#contrast-with-sealed-sender-v2
+/// [Wire Format]: #wire-format
+///
+/// # Overview
+/// The high-level steps of this process are summarized below:
+/// 1. Generate a series of random bytes.
+/// 2. Derive an ephemeral key pair from (1).
+/// 3. *Once per recipient:* Encrypt (1) using a shared secret derived from the private ephemeral
+///    key (2) and the recipient's public identity key.
+/// 4. *Once per recipient:* Add an authentication tag for (3) using a secret derived from the
+///    sender's private identity key and the recipient's public identity key.
+/// 5. Generate a symmetric key from (1) and use it to symmetrically encrypt the underlying
+///    [UnidentifiedSenderMessageContent] via [AEAD encryption]. *This step is only performed once
+///    per message, regardless of the number of recipients.*
+/// 6. Send the public ephemeral key (2), the encrypted random bytes (3), and the authentication tag
+///    (4) to the server, along with the encrypted message (5).
```

Comment:
"authentication tags" should be plural, since there's one per recipient. (I guess we can't double-pluralize "random bytes" though.)

Reply:
I've used the phrasing "the sequence of encrypted random bytes and authentication tags" to clarify this concisely here!